### ISSUE TITLE
Threaded builds/releases

### ISSUE THREAD

**Issue by maxnbk** [nerdvegas/rez#507]
It would be excellent to be able to multi-thread rez-build's per variant. I often find myself performing builds that spend a lot of time in a single-threaded linking step and could benefit by being able to rez-build or rez-release a series of variants in a threaded fashion, where builds themselves may not be able to thread well. I would expect to find it useful to be able to declare a maximum number of threads and pop unbuilt variants off the stack as a thread becomes available, to do a an "all variants in a thread all at once" type of build, and possibly to have a stagger parameter, because sometimes I have an IO-intensive first-step that might fail if I try to make the IO request 16 times at the same time.

**Comment by bfloch** [comment]
One way to achieve this is with a build process.
We've implemented a ssh-build process that will build a different variant on a different host.
Similarly you could spawn off different local builds without wait.
Why multiple threads and not simply multiple processes and dodge the GIL?

It is possible you will hit problems similar to ours:
https://github.com/nerdvegas/rez/issues/536

And one, not logged yet where the package.py is not being written atomically, so a race condition happens and only one variant is written on simultaneous updates.

**Comment by maxnbk** [comment]
Yes, perhaps agreed, though your issues with the lock files were something I had planned to suggest in the event of support for this. Perhaps there needs to be a general package-level lock and variant-level lock, to avoid race condition? package.py only written when the variant-level locks disappear, prior to the package level lock disappearing, or something along those lines.
multi-proc-ing is a good idea, but I'd still prefer it be something that the rez-build process can handle itself, as rewriting rez' cmake invocation can be awkward at times.

**Comment by bfloch** [comment]
rez build "process" plugins are not the same as rez build "system" (cmake, make etc).
The process does not know about cmake or any other build system.
Take the default 'local' build process as an sample:
https://github.com/nerdvegas/rez/blob/master/src/rezplugins/build_process/local.py

In our custom ssh process all I need to do is to filter the --process and --variants if they have been specified, or add them if available. The rest is still a rez-release with the same arguments. It does not know about cmake or any other build system.

So in our case we would `rez-release --process ssh --variants 0, 1, 2 --any-other-tag`
which would ssh into available hosts and call:
`rez-release --variants 0 --any-other-tag`
`rez-release --variants 1 --any-other-tag`
`rez-release --variants 2 --any-other-tag`

I think it is the perfect mechanism for what you try to achieve.

I agree that we need a better locking strategy. I am curious to hear Allan's opinion.
